gflownet/envs/ctorus.py

```diff
@@ -260,20 +402,74 @@ def step(
-        # If the number of actions is equal to maximum trajectory length
+        # If only possible action is EOS (number of actions is equal to maximum
+        # trajectory length), then force EOS.
         elif self.n_actions == self.length_traj:
             self.done = True
             self.n_actions += 1
             return self.state, self.eos, True
-        # If action is eos, then it is invalid
+        # If anyway action is EOS, then it is invalid
         elif action == self.eos:
+            # [DEBUG] this point should never be reached. Consider removing this elif.
+            print(
+                "An EOS action has been sampled forward even though it is not allowed"
+            )
+            breakpoint()
             return self.state, action, False
         # Otherwise perform action
         else:
+            self._step(action, backward=False)
+            return self.state, action, True
+
+    def step_backwards(
+        self, action: Tuple[float], skip_mask_check: bool = False
+    ) -> Tuple[List[float], Tuple[int, float], bool]:
+        """
+        Executes backward step given an action.
+
+        See: _step().
+
+        Args
+        ----
+        action : tuple
+            Action to be executed. An action is a vector where the value at position d
+            indicates the increment in the angle at dimension d.
+
+        skip_mask_check : bool
+            Ignored because the action space space is fully continuous, therefore there
+            is nothing to check.
+
+        Returns
+        -------
+        self.state : list
+            The sequence after executing the action
+
+        action : int
+            Action executed
+
+        valid : bool
+            False, if the action is not allowed for the current state, e.g. stop at the
+            root state
+        """
+        # If self.done is True, set done to False, increment n_actions and return same
+        # state
+        if self.done:
+            assert action == self.eos
+            self.done = False
             self.n_actions += 1
-            for dim, angle in enumerate(action):
-                self.state[int(dim)] += angle
-                self.state[int(dim)] = self.state[int(dim)] % (2 * np.pi)
-                self.state[-1] = self.n_actions
             return self.state, action, True
+        # If done is not True but action is EOS, then it is invalid
+        elif action == self.eos:
+            # [DEBUG] this point should never be reached. Consider removing this elif.
+            print(
+                "An EOS action has been sampled backward even though it is not allowed"
+            )
+            breakpoint()
+            return self.state, action, False
+        # Otherwise perform action
+        else:
+            self._step(action, backward=True)
+            return self.state, action, True
+
+    def get_max_traj_length(self):
+        return int(self.length_traj)
```

Comment:
Minor, but I am never sure, should it be `self.length_traj` or `self.length_traj + 1` (since there is also EOS)? Also, why does this need to be converted to `int`?

Reply:
Nice catch! I had actually fixed this [here](https://github.com/alexhernandezgarcia/gflownet/pull/193/commits/c31ea39466f1d4bea6780c158c9b50df91a2a48a) but had forgotten to push, sorry.

Reply:
And why the conversion to `int`?

Reply:
I don't remember tbh.